README.md
# YugabyteDB Managed - Allow List Management

[![Build](https://github.com/yogendra/ybm-allow-list-management/actions/workflows/build.yaml/badge.svg)](https://github.com/yogendra/ybm-allow-list-management/actions/workflows/build.yaml)

This module can help in dynamically updating allow list in case of serverless /
PaaS environments where egress IPs are dynamic.

This is developed particularly to support accessing
[YugabyteDB Managed](https://cloud.yugabte.com) instance from [fly.to]. This uses
[YBM APIs](https://api-docs.yugabyte.com/docs/managed-apis) to update the
network allow list at runtime

## Features

1. Create versioned allow list
2. Easy to override or configure manually, just follow the naming conventions
3. Create new version on each run
4. Skips version creations on duplicate updates
5. Dis-associated old version from cluster.
6. Keeps existing association from older version

## Example

```js
// Make sure YBM_API_KEY. YBM_PROJECT_ID and YBM_ACCOUNT_ID are defined in the environment
import { update } from "ybm-allow-list-management";
let prefix = "myapp-allow-list"
// Find your apps egress IP
let egressIp = await fetch("https://ifconfig.me/ip",)
  .then(res => {
    return res.text();
  })
  .then(ipAddress => `${ipAddress}/32`);
// Get it from the YBM console address bar
let clusterId = '<cluster-uuid>'

console.log("Egress IP: " + egressIp);

let allowList = await update(prefix, egressIp, clusterId);
console.log("Allow list name " + allowList.spec.name);
```

Code above will go create an allow list with prefix `myapp-allow-list`. If there is any existing list matching
`myapp-allow-list--v#` format, it will create new list by addin `+1` to version. It will copy over existing
ip addresses and cluster ids from older version.

## Environment Variables

Following environment variables are used by this module

- `YBM_API_KEY` - (**Required**) YugabyteDB Managed - Api Key
- `YBM_ACCOUNT_ID` - (**Required**) YugabyteDB Managed - Account ID
- `YBM_PROJECT_ID` - (**Required**) YugabyteDB Managed - Project ID
- `YBM_ENDPOINT` - (_Optional_) YugabyteDB Managed - Endpoint URL (Default: `https://cloud.yugabyte.com/api`)
- `YBM_MAX_RETRY` - (_Optional_) Maximum retry for api requests (Default : `30`)
- `YBM_RETRY_INTERVAL` - (_Optional_) Internal (seconds) between api requests (Default : `2`)
- `YBM_CLUSTER_ID` - (_Test Only_) YugabyteDB Managed - Cluster ID
- `YBM_ALLOW_LIST_PREFIX` - (_Test Only_) Prefix for allow list name
- `NODE_ENV` - (_Test Only_) Set to `development` for debug logs

## Develop

_TBA_


## Get in Touch

Get in touch via [YugabyteDB Community Slack](https://yugabyte.com/slack)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## 1.0.42 - 2023-02-14

## 1.0.41 - 2023-02-14

## 1.0.40 - 2023-02-14

## 1.0.39 - 2023-02-14

## 1.0.38 - 2023-02-14

## 1.0.37 - 2023-02-14

## 1.0.36 - 2023-02-14

## 1.0.35 - 2023-02-14

## 1.0.34 - 2023-02-14

## 1.0.33 - 2023-02-14

## 1.0.32 - 2023-02-14

## 1.0.31 - 2023-01-26

## 1.0.30 - 2023-01-26

## 1.0.29 - 2023-01-26

## 1.0.28 - 2023-01-26

## 1.0.26 - 2023-01-26
### Added
- Unit testing

### Fixed
- Timing issue

## 1.0.17 - 2023-01-26
### Added
- Basic functions - update
- E2E testing
- Release automation

### Fixed
- Testing


index.test.js
import { jest, expect, describe, test, beforeAll, afterAll } from '@jest/globals'

/**
 * Preparation
 * 1. Create a cluster
 * 2. Set Env vars : YBM_API_KEY, YBM_PROJECT_ID, YBM_ACCOUNT_ID, YBM_CLUSTER_ID
 */

import { update } from './index.js'
import YBMClient from './ybm-client.js'
import { sleep, debug } from './util.js'

jest.setTimeout(600000)

jest.setTimeout(600000)

const testConfig = {
  prefix: process.env.YBM_ALLOWLIST_PREFIX,
  clusterId: process.env.YBM_CLUSTER_ID,
  accountId: process.env.YBM_ACCOUNT_ID,
  projectId: process.env.YBM_PROJECT_ID,
  apiKey: process.env.YBM_API_KEY,
  endpoint: process.env.YBM_ENDPOINT,
  maxRetry: process.env.YBM_MAX_RETRY || 30,
  retryInterval: process.env.YBM_RETRY_INTERVAL || 2
}

describe('Updte Allow List', () => {
  beforeAll(async () => {
    await deleteTestData()
  })
  afterAll(async () => {
    await deleteTestData()
  })

  test('clean test start', () => {
    expect(true).toBeTruthy()
  })
  let u1, u2, u3, u4

  test('Create new', async () => {
    // Create new list v1
    u1 = await update(testConfig.prefix, '127.0.0.1/32', testConfig.clusterId)
    debug(JSON.stringify(u1))
    expect(u1.info.id).toBeDefined()
    expect(u1.info.cluster_ids.includes(testConfig.clusterId)).toBeTruthy()
    expect(u1.spec.name).toEqual(testConfig.prefix + '--v1')
    expect(u1.spec.allow_list.includes('127.0.0.1/32'))
  })

  test('Update IP', async () => {
    // Update list, new version, copy over associations
    u2 = await update(testConfig.prefix, '127.0.0.2/32')
    expect(u2.info.id).toBeDefined()
    expect(u2.info.cluster_ids.includes(testConfig.clusterId)).toBeTruthy()
    expect(u2.spec.name).toEqual(testConfig.prefix + '--v2')
    expect(u2.spec.allow_list.includes('127.0.0.1/32'))
    expect(u2.spec.allow_list.includes('127.0.0.2/32'))
  })

  test('Update duplicate IP (no-op)', async () => {
    // Skip duplicate updates
    u3 = await update(testConfig.prefix, '127.0.0.2/32')
    expect(u3.info.id).toBeDefined()
    expect(u3.info.id).toEqual(u2.info.id)
    expect(u3.info.cluster_ids.includes(testConfig.clusterId)).toBeTruthy()
    expect(u3.spec.name).toEqual(u2.spec.name)
    expect(u3.spec.allow_list.includes('127.0.0.1/32'))
    expect(u3.spec.allow_list.includes('127.0.0.2/32'))
  })

  test('Update new IP', async () => {
    //
    u4 = await update(testConfig.prefix, '127.0.0.3/32')
    expect(u4.info.id !== u3.info.id).toBeTruthy()
    expect(u4.info.cluster_ids.includes(testConfig.clusterId)).toBeTruthy()
    expect(u4.spec.name).toEqual(testConfig.prefix + '--v3')
    expect(u4.spec.allow_list.includes('127.0.0.1/32'))
    expect(u4.spec.allow_list.includes('127.0.0.2/32'))
    expect(u4.spec.allow_list.includes('127.0.0.3/32'))
  })
})

async function deleteTestData () {
  const ybm = new YBMClient(testConfig.apiKey, testConfig.accountId, testConfig.projectId, testConfig.endpoint)

  console.info('Get allow lists for an existing cluster')
  let testAllowLists = await ybm.get('/allow-lists', { limit: 1000, order_by: 'name' })
  testAllowLists = testAllowLists.data.filter(x => x.spec.name.startsWith(testConfig.prefix))
  const testAllowListIds = testAllowLists.map(x => x.info.id).sort()
  console.info('Delete Test Data - Remove allow list associations')
  const clusterUpdates = testAllowLists
    .flatMap(x => x.info.cluster_ids)
    .map(async (clusterId) => {
      const currentList = (await ybm.get(`/clusters/${clusterId}/allow-lists`)).data
      const updateAllowListIds = currentList
        .filter(x => !testAllowListIds.includes(x.info.id))
        .map(x => x.info.id)
      let retry = testConfig.maxRetry
      while (--retry > 0) {
        const update = await ybm.put(`/clusters/${clusterId}/allow-lists`, updateAllowListIds)
        if (update.error) {
          console.warn(`Update List Failed: ${update.error.status} - ${update.error.detail}`)
        }
        if (update.data) {
          break
        }
        await sleep(testConfig.retryInterval)
      }
      retry = testConfig.maxRetry
      while (--retry > 0) {
        const update = await ybm.get(`/clusters/${clusterId}/allow-lists`)
        const updateIds = update.data.map(x => x.info.id)
        const updateCompleted = testAllowListIds.every(x => !updateIds.includes(x))
        if (updateCompleted) {
          break
        }
        await sleep(testConfig.retryInterval)
      }
      return true
    })

  await Promise.all(clusterUpdates)

  console.info('Delete tests data - allow lists')
  const deleteResponses = testAllowListIds.map(async id => {
    return ybm.delete(`/allow-lists/${id}`)
  })
  await Promise.all(deleteResponses)

  console.info('Test Data Cleaned Up')
}


test/index.js
/* eslint-disable no-unused-expressions */
/**
 * Preparation
 * 1. Create a cluster
 * 2. Set Env vars : YBM_API_KEY, YBM_PROJECT_ID, YBM_ACCOUNT_ID, YBM_CLUSTER_ID
 */

const { describe, before, after, it } = require('mocha')
const { expect, assert } = require('chai')

const update = require('../index').update
const { debug } = require('../util')
const testConfig = require('./test-config')
const deleteTestData = require('./test-data')

before('Before Tests - Delete All Test Data', async function () {
  this.timeout(60000)
  await deleteTestData()
})
after('After Tests - Delete All Test Data', async function () {
  this.timeout(60000)
  await deleteTestData()
})
describe('Allow List', function () {
  this.timeout(60000)
  it('clean test start', function () {
    assert.equal(true, true, 'test started')
  })

  let u1, u2, u3, u4

  it('Create new', async function () {
    // Create new list v1
    u1 = await update(testConfig.prefix, '127.0.0.1/32', testConfig.clusterId)
    debug(JSON.stringify(u1))
    expect(u1.info.id).to.exist
    expect(u1.info.cluster_ids).to.include(testConfig.clusterId)
    expect(u1.spec.name).to.equal(testConfig.prefix + '--v1')
    expect(u1.spec.allow_list).to.include('127.0.0.1/32')
  })

  it('Update IP', async function () {
  // Update list, new version, copy over associations
    u2 = await update(testConfig.prefix, '127.0.0.2/32')
    expect(u2.info.id).to.exist
    expect(u2.info.cluster_ids).to.include(testConfig.clusterId)
    expect(u2.spec.name).to.equal(testConfig.prefix + '--v2')
    expect(u2.spec.allow_list).to.include('127.0.0.1/32')
    expect(u2.spec.allow_list).to.include('127.0.0.2/32')
  })

  it('Update duplicate IP (no-op)', async function () {
  // Skip duplicate updates
    u3 = await update(testConfig.prefix, '127.0.0.2/32')
    expect(u3.info.id).to.exist
    expect(u3.info.id).to.equal(u2.info.id)
    expect(u3.info.cluster_ids).to.include(testConfig.clusterId)
    expect(u3.spec.name).to.equal(u2.spec.name)
    expect(u3.spec.allow_list).to.include('127.0.0.1/32')
    expect(u3.spec.allow_list).to.include('127.0.0.2/32')
  })

  it('Update new IP', async function () {
  //
    u4 = await update(testConfig.prefix, '127.0.0.3/32')
    expect(u4.info.id).to.not.equal(u3.info.id)
    expect(u4.info.cluster_ids).to.include(testConfig.clusterId)
    expect(u4.spec.name).to.equal(testConfig.prefix + '--v3')
    expect(u4.spec.allow_list).to.include('127.0.0.1/32')
    expect(u4.spec.allow_list).to.include('127.0.0.2/32')
    expect(u4.spec.allow_list).to.include('127.0.0.3/32')
  })
})


test/test-data.js
const { sleep } = require('../util')
const YBMClient = require('../ybm-client')
const testConfig = require('./test-config')

async function deleteTestData () {
  const ybm = new YBMClient(testConfig.apiKey, testConfig.accountId, testConfig.projectId, testConfig.endpoint)

  console.info('TEST DATA Setup :: Get allow lists for an existing cluster')
  let testAllowLists = await ybm.get('/allow-lists', { limit: 1000, order_by: 'name' })
  testAllowLists = testAllowLists.data.filter(x => x.spec.name.startsWith(testConfig.prefix))
  const testAllowListIds = testAllowLists.map(x => x.info.id).sort()
  console.info('TEST DATA Setup :: Delete Test Data - Remove allow list associations')
  const clusterUpdates = testAllowLists
    .flatMap(x => x.info.cluster_ids)
    .map(async (clusterId) => {
      const currentList = (await ybm.get(`/clusters/${clusterId}/allow-lists`)).data
      const updateAllowListIds = currentList
        .filter(x => !testAllowListIds.includes(x.info.id))
        .map(x => x.info.id)
      let retry = testConfig.maxRetry
      while (--retry > 0) {
        const update = await ybm.put(`/clusters/${clusterId}/allow-lists`, updateAllowListIds)
        if (update.error) {
          console.warn(`TEST DATA Setup :: Update List Failed: ${update.error.status} - ${update.error.detail}`)
        }
        if (update.data) {
          break
        }
        await sleep(testConfig.retryInterval)
      }
      retry = testConfig.maxRetry
      while (--retry > 0) {
        const update = await ybm.get(`/clusters/${clusterId}/allow-lists`)
        const updateIds = update.data.map(x => x.info.id)
        const updateCompleted = testAllowListIds.every(x => !updateIds.includes(x))
        if (updateCompleted) {
          break
        }
        await sleep(testConfig.retryInterval)
      }
      return true
    })

  await Promise.all(clusterUpdates)

  console.info('TEST DATA Setup :: Delete tests data - allow lists')
  const deleteResponses = testAllowListIds.map(async id => {
    return ybm.delete(`/allow-lists/${id}`)
  })
  await Promise.all(deleteResponses)

  console.info('TEST DATA Setup :: Test Data Cleaned Up')
}

module.exports = deleteTestData


ybm-client.js
const debug = require('./util').debug

class YBMClient {
  constructor (apiKey, accountId, projectId, endpoint) {
    const ep = endpoint || 'https://cloud.yugabyte.com/api'
    this.apiRoot = `${ep}/public/v1/accounts/${accountId}/projects/${projectId}`
    this.commonHeaders = {
      Authorization: `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
      Accept: 'application/json'
    }
  }

  async request (method, path, param = null, body = null, headers = null) {
    const qs =
      param !== null ? '?' + new URLSearchParams(param).toString() : ''
    const url = `${this.apiRoot}${path}${qs}`
    const options = {
      headers: { ...this.commonHeaders, ...headers },
      method,
      body: body !== null ? JSON.stringify(body) : null
    }
    debug({ _tag: 'info', _method: 'ybm.request', url, options })
    return fetch(url, options)
      .catch((x) => {
        console.error(`YBM API: ${path}, error: ${x}}`)
        throw new Error(x)
      })
      .then(async (response) => {
        const contentType = response.headers.get('content-type')
        if (contentType && contentType.indexOf('application/json') !== -1) {
          return response.json()
        } else {
          const bodyText = await response.text()
          if (method === 'DELETE' && bodyText === '') {
            return {}
          }
          const errorInfo = {
            req: { url, method, param, body, headers },
            res: {
              type: contentType,
              body: bodyText
            }
          }
          debug({
            _tag: 'error',
            _method: 'ybm.request',
            errorInfo
          })
          throw new Error('Non JSON Response')
        }
      })
  }

  async get (path, params, headers) {
    return this.request('GET', path, params, null, headers)
  }

  async post (path, body, headers) {
    return this.request('POST', path, null, body, headers)
  }

  async put (path, body, headers) {
    return this.request('PUT', path, null, body, headers)
  }

  async delete (path, params, headers) {
    return this.request('DELETE', path, params, null, headers)
  }
}

module.exports = YBMClient


index.js
const YBMClient = require('./ybm-client.js')
const { debug, sleep } = require('./util.js')

const config = {
  accountId: process.env.YBM_ACCOUNT_ID,
  projectId: process.env.YBM_PROJECT_ID,
  apiKey: process.env.YBM_API_KEY,
  maxRetry: process.env.YBM_MAX_RETRY || 30,
  retryInterval: process.env.YBM_RETRY_INTERVAL || 2
}
assert(config.apiKey, 'API key not set (YBM_API_KEY)')
assert(config.accountId, 'Account ID not set (YBM_ACCOUNT_ID)')
assert(config.accountId, 'Project ID not set (YBM_PROJECT_ID)')

const ybm = new YBMClient(config.apiKey, config.accountId, config.projectId)

/*
Basic Idea:
Find a lists matching a <prefix>-v<version> expression
Get the latest allow list and extract
1. version
2. associated cluster ids
3. cidr list

Prepare new allow list def
1. name = prefix + c + (version+1)
2. cidr = new cidrs + cidr list form latest allow list
3. description = "Allow list <comma-separated-cidr-list>"

Update allow list association for each of the cluster associated with latest allow list
1. get all the allow list ids from cluster
2. remove all the old allow list ids
3. add new allow list id

*/
async function update (prefix, cidrOrIpList, clusterId = null) {
  const prefixRegex = new RegExp(`^${prefix}--v(\\d+)$`)

  let nextNumber = 1
  let newAllowList = Array.isArray(cidrOrIpList)
    ? cidrOrIpList
    : [cidrOrIpList]
  let latestAllowList
  let clusterIds = clusterId !== null ? [clusterId] : []
  debug(
    `Update allow list with prefix ${prefix} to include cidr ${newAllowList.join(
      ','
    )}`
  )

  let existingLists = await getAllowListByName(prefixRegex)

  const existingListIds = existingLists.map((x) => x.info.id).sort()
  if (Array.isArray(existingLists) && existingLists.length > 0) {
    debug('Found existing allow list with prefix')
    existingLists = existingLists.sort((a, b) => {
      // Sort descending based on index
      const indexA = +prefixRegex.exec(a.spec.name)[1]
      const indexB = +prefixRegex.exec(b.spec.name)[1]
      return indexB - indexA
    })

    latestAllowList = existingLists[0]
    debug(
      `Latest allow list: id: ${latestAllowList.info.id}, name: ${latestAllowList.spec.name}`
    )
    const clusterIdPresent =
      clusterId === null ||
      latestAllowList.info.cluster_ids.includes(clusterId)
    const allCidrsPresent = newAllowList.every((x) =>
      latestAllowList.spec.allow_list.includes(x)
    )
    if (clusterIdPresent && allCidrsPresent) {
      debug(
        `Skipping updated all as cluster (${clusterId}) and CIDRs (${newAllowList.join(
          ','
        )}) present`
      )
      return latestAllowList
    }
    clusterIds = clusterIds.concat(latestAllowList.info.cluster_ids)
    nextNumber = ++prefixRegex.exec(latestAllowList.spec.name)[1]
    newAllowList = newAllowList
      .concat(latestAllowList.spec.allow_list)
      .sort()
      .filter((v, i, a) => a.indexOf(v) === i) // uniq element trick
  }

  const name = `${prefix}--v${nextNumber}`
  const description = `Allow list for ${newAllowList.join(',')}`

  debug(
    `Create a new allow list: name: ${name}, description: ${description}, allow list: ${newAllowList.join(
      ','
    )}`
  )
  const updatedList = await createAllowList(name, description, newAllowList)
  const updatedListId = updatedList.info.id

  debug(`Created allow list: id: ${updatedListId}`)
  const updatedClusterAllowLists = await Promise.all(
    clusterIds.map(async (id) => {
      const currentClusterAllowListIds = await getClusterAllowListIds(id)
      const updatedClusterAllowListIds = currentClusterAllowListIds.filter(
        (x) => existingListIds.indexOf(x) === -1
      )
      updatedClusterAllowListIds.push(updatedListId)
      debug(
        `Update allow list for cluster (${id}) to (${updatedClusterAllowListIds.join(
          ','
        )})`
      )
      return await updateClusterAllowLists(id, updatedClusterAllowListIds)
    })
  )
  debug(
    `Clusters (${clusterIds.join(',')} (responses: ${
      updatedClusterAllowLists.length
    }))`
  )
  return await getAllowList(updatedListId)
}

async function getAllowListByName (regex) {
  const path = '/allow-lists'
  const params = {
    order_by: 'name',
    limit: '1000',
    is_auto_created: 'false'
  }

  const response = await ybm.get(path, params)
  const filteredList = response.data.filter((x) => regex.test(x.spec.name))
  debug({
    _tag: 'debug',
    _method: 'getAllowListByName',
    arguments,
    filteredList,
    ...response
  })
  return filteredList
}

async function getClusterAllowListIds (id) {
  debug(`getClusterAllowListIds - Fetch allow list ids for cluster (${id})`)
  const path = `/clusters/${id}/allow-lists`
  const response = await ybm.get(path)
  if (response.error) {
    console.warn(
      'Encountered error in get allow lists: ' + errorDetails(response)
    )
    return null
  }
  return response.data.map((x) => x.info.id).sort()
}

async function createAllowList (name, description, cidrOrIpList) {
  const allowList = Array.isArray(cidrOrIpList) ? cidrOrIpList : [cidrOrIpList]
  debug(
    `createAllowList - name: ${name}, desc: ${description}, allowList: ${allowList}`
  )
  const path = '/allow-lists'
  const body = {
    name,
    description,
    allow_list: allowList
  }
  const response = await ybm.post(path, body)
  if (
    response &&
    response.data &&
    response.data.spec &&
    response.data.info &&
    response.data.info.id
  ) {
    return response.data
  }
  debug({
    _tag: 'error',
    _method: 'createAllowList',
    _msg: 'Error to create list',
    ...response
  })
  throw new Error('Failed to create allow list')
}

async function updateClusterAllowLists (clusterId, allowListIds) {
  const path = `/clusters/${clusterId}/allow-lists`
  const param = allowListIds.sort()
  let retry = config.maxRetry

  while (--retry > 0) {
    const response = await ybm.put(path, param)
    if (response.data) {
      break
    }
    if (response.error) {
      console.warn('Encountered error in update :' + errorDetails(response))
    }
    await sleep(config.retryInterval)
  }
  retry = config.maxRetry
  while (--retry > 0) {
    const ids = await getClusterAllowListIds(clusterId)
    if (ids && allowListIds.every((x) => ids.includes(x))) {
      return ids
    }
    debug('Cluster allow list update not completed :' + ids.join(','))

    await sleep(config.retryInterval)
  }
  debug({
    _tag: 'error',
    _method: 'updateClusterAllowLists',
    _msg: 'updateClusterAllowLists'
  })
  throw new Error(
    `Failed update:  cluster(${clusterId}), allow lists:(${param})`
  )
}

async function getAllowList (id) {
  const path = `/allow-lists/${id}`
  const response = await ybm.get(path, {})
  return response.data
}
function assert (x, error) {
  if (typeof x === 'undefined' || x === null) {
    throw new Error(error)
  }
}
function errorDetails (res) {
  if (res && res.error && res.error.status && res.error.detail) {
    return `${res.error.status} - ${res.error.detail}`
  } else {
    return JSON.stringify(res)
  }
}

module.exports = { update }
